stage/LibUsbDotNet/MonoLibUsb/Profile/MonoUsbProfileList.cs

```diff
@@ -55,50 +55,52 @@ private void SetDiscovered(bool discovered)
-            {
-                MonoUsbProfile newProfile = newList.mList[iNewProfile];
-                int iFoundOldIndex;
-                if ((iFoundOldIndex = mList.IndexOf(newProfile)) == -1)
-                {
-                    //Console.WriteLine("DeviceDiscovery: Added: {0}", newProfile.ProfileHandle.DangerousGetHandle());
-                    newProfile.mDiscovered = true;
-                    mList.Add(newProfile);
-                    FireAddRemove(newProfile, AddRemoveType.Added);
-                }
-                else
-                {
-                    //Console.WriteLine("DeviceDiscovery: Unchanged: Orig:{0} New:{1}", mList[iFoundOldIndex].ProfileHandle.DangerousGetHandle(), newProfile.ProfileHandle.DangerousGetHandle());
-                    mList[iFoundOldIndex].mDiscovered = true;
-                    newProfile.mDiscovered = false;
-                   
-                }
-            }
-
-            newList.mList.RemoveAll(FindDiscoveredFn);
-            newList.Close();
-
-            foreach (MonoUsbProfile deviceProfile in mList.ToList())
-            {
-                if (!deviceProfile.mDiscovered)
-                {
-                    // Close Unplugged device profiles.
-                    //Console.WriteLine("DeviceDiscovery: Removed: {0}", deviceProfile.ProfileHandle.DangerousGetHandle());
-                    FireAddRemove(deviceProfile, AddRemoveType.Removed);
-                    deviceProfile.Close();
-                }
-            }
-
-            // Remove Unplugged device profiles.
-            mList.RemoveAll(FindUnDiscoveredFn);
+        }
+
+        private void syncWith(MonoUsbProfileList newList)
+        {
+            SetDiscovered(false);
+            newList.SetDiscovered(true);
+            int iNewProfiles = newList.mList.Count;
+            for (int iNewProfile = 0; iNewProfile < iNewProfiles; iNewProfile++)
+            {
+                MonoUsbProfile newProfile = newList.mList[iNewProfile];
+                int iFoundOldIndex;
+                if ((iFoundOldIndex = mList.IndexOf(newProfile)) == -1)
+                {
+                    newProfile.mDiscovered = true;
+                    mList.Add(newProfile);
+                    FireAddRemove(newProfile, AddRemoveType.Added);
+                }
+                else
+                {
+                    if (newProfile.ProfileHandle.DangerousGetHandle() != mList[iFoundOldIndex].ProfileHandle.DangerousGetHandle())
+                    {
+                        newProfile.mDiscovered = true;
+                        mList.Add(newProfile);
+                        FireAddRemove(newProfile, AddRemoveType.Added);
+                        mList[iFoundOldIndex].mDiscovered = false;
+                    }
+                    else
+                    {
+                        mList[iFoundOldIndex].mDiscovered = true;
+                        newProfile.mDiscovered = false;
+                    }
+                }
```

Comment:
This block could be written using `else if` for readability and to reduce nesting

Reply:
Less  diff from the original version this  way though

Reply:
Updated with continues for even better readability (in a C coder's eyes at least)